core/src/main/java/com/sun/mail/util/WriteTimeoutSocket.java

```diff
@@ -398,17 +426,26 @@ public synchronized void write(byte[] bs, int off, int len)
 	}
 
 	// Implement timeout with a scheduled task
-	ScheduledFuture<Object> sf = null;
+	ScheduledFuture<String> sf = null;
 	try {
-	    try {
-		if (timeout > 0)
-		    sf = ses.schedule(timeoutTask,
-					timeout, TimeUnit.MILLISECONDS);
-	    } catch (RejectedExecutionException ex) {
-		// ignore it; Executor was shut down by another thread,
-		// the following write should fail with IOException
-	    }
-	    os.write(bs, off, len);
+        try {
+            if (timeout > 0)
+                sf = ses.schedule(timeoutTask,
+                    timeout, TimeUnit.MILLISECONDS);
+        } catch (RejectedExecutionException ex) {
+            if (!socket.isClosed()) {
+                IOException ioex = new IOException("Write aborted due to timeout not enforced");
+                ioex.initCause(ex);
+                throw ioex;
+            }
+        }
+
+        try {
+            os.write(bs, off, len);
+        } catch (IOException e) {
+            e = sf == null ? e : new IOException(getTimeoutTaskResult(sf), e);
+            throw e;
+        }
 	} finally {
 	    if (sf != null)
 		sf.cancel(true);
```

Comment:
I just noticed bug here.  We have to make sure that cancel is called before we get the result or we are going to end up blocking for full timeout every call.

We have one of two options here:

1.  Add another try block so the finally block can run before we catch IOException.  Then getTimeoutTaskResult should add an if condition to not call sf.get when sf.isCancelled() is true.
2.  Call sf.cancel(true) in getTimeoutTaskResult and if the result is false proceed with sf.get.

Reply:
fixed using the 2nd suggestion

Reply:
I will also try to add a test for checking exceptions caught in handleTimeoutTaskResult.
Btw, do you mind if I squash all commits of the feature into one?

Reply:
Squashing the commits in this branch is fine.